### PR TITLE
#54 - Support for absent values when pulling a layer

### PR DESCRIPTION
**Pull request by olegmoz**
Part of #54
Changed AstoBlobs.blob() method to return no content when it is absent.
Pull layer request returns 404 when content is missing.

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#78](https://github.com/artipie/docker-adapter/issues/78) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/227d12de-0496-41db-8cf9-cb91fed7de16); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by olegmoz** [comment]
@Vatavuk ping

File: src/main/java/com/artipie/docker/BlobStore.java [hunk 1]
```diff
@@ -38,7 +39,7 @@
      * @param digest Blob digest
      * @return Async publisher output
      */
-    CompletableFuture<Content> blob(Digest digest);
+    CompletableFuture<Optional<Content>> blob(Digest digest);
```
Comment (Vatavuk) [inline src/main/java/com/artipie/docker/BlobStore.java:42]: @olegmoz can we check if content is present  by inspecting `Content.size().isPresent()`?

File: src/main/java/com/artipie/docker/asto/AstoBlobs.java [hunk 2]
```diff
@@ -78,11 +80,18 @@ public AstoBlobs(final Storage asto) {
...
-            )
+    public CompletableFuture<Optional<Content>> blob(final Digest digest) {
+        final Key key = new Key.From(RegistryRoot.V2, new BlobRef(digest).string(), "data");
+        return this.asto.exists(key).thenCompose(
+            exists -> {
+                final CompletionStage<Optional<Content>> stage;
+                if (exists) {
+                    stage = this.asto.value(key).thenApply(Optional::of);
+                } else {
+                    stage = CompletableFuture.completedStage(Optional.empty());
```
Comment (Vatavuk) [inline src/main/java/com/artipie/docker/asto/AstoBlobs.java:91]: @olegmoz can we create an empty implementation of `Content` that will return `Optional.empty()` in `Content.size()` method?

File: src/main/java/com/artipie/docker/http/PullLayer.java [hunk 3]
```diff
@@ -81,11 +82,15 @@ public Response response(
             );
         }
         final Digest digest = new Digest.FromString(matcher.group("digest"));
-        return connection -> this.docker.blobStore().blob(digest).thenCompose(
-            layer -> new RsWithBody(
-                new RsWithStatus(RsStatus.OK),
-                layer
-            ).send(connection)
+        return new AsyncResponse(
+            this.docker.blobStore().blob(digest).thenApply(
```
Comment (Vatavuk) [inline src/main/java/com/artipie/docker/http/PullLayer.java:86]: @olegmoz here we can check if `size` is present.

File: src/main/java/com/artipie/docker/BlobStore.java (hunk 1 above)
Comment (olegmoz) [reply src/main/java/com/artipie/docker/BlobStore.java:42]: @Vatavuk no unfortunately we cannot. By contract `Storage.value()` throws exception if value does not exist.

File: src/main/java/com/artipie/docker/asto/AstoBlobs.java (hunk 2 above)
Comment (olegmoz) [reply src/main/java/com/artipie/docker/asto/AstoBlobs.java:91]: @Vatavuk we can, but I see not point in doing so. Empty value and absent value are considered to be different (think of file system as example).

File: src/main/java/com/artipie/docker/http/PullLayer.java (hunk 3 above)
Comment (olegmoz) [reply src/main/java/com/artipie/docker/http/PullLayer.java:86]: @Vatavuk please see above comment, existence cannot be checked by looking at content size

File: src/main/java/com/artipie/docker/asto/AstoBlobs.java (hunk 2 above)
Comment (Vatavuk) [reply src/main/java/com/artipie/docker/asto/AstoBlobs.java:91]: @olegmoz ok I get it, if size is not present than we have an empty content but its not missing. Then my comments can't be applied.

**Comment by 0crat** [comment]
@vatavuk[/z](https://www.0crat.com/u/vatavuk) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by Vatavuk** [comment]
@g4s8 please remove this issue from the scope

**Comment by g4s8** [comment]
@0crat out

**Comment by 0crat** [comment]
> @0crat out ([here](https://github.com/artipie/docker-adapter/issues/78#issuecomment-618362463))

@g4s8 There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
PID: 4@c08ec908-6891-4027-9aad-79415e47ffee, thread: PQ-CT2E6TK9B
com.jcabi.github.RtValuePagination$Items[181] java.lang.IllegalStateException: java.io.IOException: failed after 4 attempts
com.jcabi.http.wire.RetryWire[95] java.io.IOException: failed after 4 attempts
```

[0.53.11](https://github.com/zerocracy/farm): CID: [6f0fb9f0-be7d-41de-bb8a-02d8d659d7e5](https://www.0crat.com/footprint/CT2E6TK9B/6f0fb9f0-be7d-41de-bb8a-02d8d659d7e5), Type: "Finish order"

**Comment by 0crat** [comment]
> @0crat out ([here](https://github.com/artipie/docker-adapter/issues/78#issuecomment-618362463))

@g4s8 There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
PID: 4@c08ec908-6891-4027-9aad-79415e47ffee, thread: PQ-CT2E6TK9B
com.zerocracy.farm.reactive.StkRuntime[76] java.lang.IllegalStateException: org.codehaus.groovy.runtime.InvokerInvocationException in com.zerocracy.stk.pm.qa.add_label_on_qa
org.codehaus.groovy.reflection.CachedMethod[105] org.codehaus.groovy.runtime.InvokerInvocationException: java.io.IOException: failed after 4 attempts
com.jcabi.http.wire.RetryWire[95] java.io.IOException: failed after 4 attempts
```

[0.53.11](https://github.com/zerocracy/farm): CID: [6c288e9c-388a-4c7e-8092-ba4847f48e50](https://www.0crat.com/footprint/CT2E6TK9B/6c288e9c-388a-4c7e-8092-ba4847f48e50), Type: "Start QA review"

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @vatavuk[/z](https://www.0crat.com/u/vatavuk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Code review was too long (8 days), architects (@g4s8) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Pull request merged**